<a href="https://colab.research.google.com/github/amurtigro041/6Kyu_Solution/blob/main/Ecommerce_Dataset_Kresna_Bangkit.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

 Proyek Analisis Data : Ecommerce Public Dataset


*   Nama : Kresna Fajri Wicaksana
*   Email : kresna.fajri@gmail.com
*   ID Dicoding : Kresna Fajri

Menentukan Pertanyaan Bisnis


1. Ada berapa jumlah pelanggan di tiap daerah/state?
2. Pembeli terbanyak ada di state mana?  
3. Apakah ada barang yang tidak sampai pada pelanggan?Jika ada, di
   negara bagian/kota mana saja?
4. Produk apa saja yang paling laris dan paling tidak laris?

# Proses Import paket


In [85]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import scipy
import seaborn

In [86]:
!pip install streamlit

In [87]:
import streamlit as st

# Data Wrangling

Gathering and Cleaning Data

In [99]:
#Akan dilakukan proses impor dan pembacaan data
customer_df = pd.read_csv("/content/customers_dataset.csv",engine ="python")
customer_df.head(10)
print(customer_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB
None


In [100]:
print("deskripsi statistik pada customer_df \n")

print(customer_df.describe(include="all"))

print(f'Jumlah NaN values pada dataframe customer_df yaitu {customer_df.isna().sum()}')
print(f'Jumlah duplikasi pada dataframe customer_df yaitu {customer_df.duplicated().sum()} ')

deskripsi statistik pada customer_df 

                             customer_id                customer_unique_id  \
count                              99441                             99441   
unique                             99441                             96096   
top     06b8999e2fba1a1fbc88172c00ba8bc7  8d50f5eadf50201ccdcedfb9e2ac8455   
freq                                   1                                17   
mean                                 NaN                               NaN   
std                                  NaN                               NaN   
min                                  NaN                               NaN   
25%                                  NaN                               NaN   
50%                                  NaN                               NaN   
75%                                  NaN                               NaN   
max                                  NaN                               NaN   

        customer_zip_cod

In [98]:
#baca 10 data teratas orderitem_df
orderitem_df.head(5)

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

In [115]:
print(f"Info orderItem_df {orderitem_df.info()} \n\n--------------Batas Suci---------------\n")

print(f'Jumlah NaN values pada dataframe orderitem_df yaitu {orderitem_df.isna().sum()}')
print(f'Jumlah duplikasi pada dataframe orderitem_df yaitu {orderitem_df.duplicated().sum()}')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB
Info orderItem_df None 

--------------Batas Suci---------------

Jumlah NaN values pada dataframe orderitem_df yaitu order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64
Jumlah duplikasi pada dataframe orderitem_df yaitu 0


In [156]:
#Lanjutkan lagi dengan order payment
orderpayment_df = pd.read_csv("/content/order_payments_dataset.csv")
print(orderpayment_df.head(5))


                           order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45  


In [155]:
#cek info orderpayment_df dan cek duplicates dan NaN value
print(f'Jumlah data terduplikasi : {orderpayment_df.duplicated().sum()}')
print(f'Jumlah NaN value : {orderpayment_df.isna().sum()}')
#Data clean
print(orderpayment_df.describe())

Jumlah data terduplikasi : 0
Jumlah NaN value : order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64
       payment_sequential  payment_installments  payment_value
count       103886.000000         103886.000000  103886.000000
mean             1.092679              2.853349     154.100380
std              0.706584              2.687051     217.494064
min              1.000000              0.000000       0.000000
25%              1.000000              1.000000      56.790000
50%              1.000000              1.000000     100.000000
75%              1.000000              4.000000     171.837500
max             29.000000             24.000000   13664.080000


In [133]:
orders_df = pd.read_csv("/content/orders_dataset.csv",engine="python")
#cetak info dataset order_df
print(f"Informasi dataset order_df{orders_df.info()}\n\n-----------------------\n\n")
print(f'Jumlah NaN values pada dataframe orders_df yaitu {orders_df.isna().sum()}')
print(f'Jumlah duplikasi pada dataframe orders_df yaitu {orders_df.duplicated().sum()}')


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB
Informasi dataset order_dfNone

-----------------------


Jumlah NaN values pada dataframe orders_df yaitu order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
ord

In [157]:
#Lakukan cleaning pada data
orders_df.dropna(subset=["order_approved_at","order_delivered_carrier_date","order_delivered_customer_date"],inplace=True)
#Cek ulang
orders_df.isna().sum()

order_id                         0
customer_id                      0
order_status                     0
order_purchase_timestamp         0
order_approved_at                0
order_delivered_carrier_date     0
order_delivered_customer_date    0
order_estimated_delivery_date    0
dtype: int64

In [117]:
print(orders_df.describe(include="all"))

                                order_id                       customer_id  \
count                              99441                             99441   
unique                             99441                             99441   
top     e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
freq                                   1                                 1   

       order_status order_purchase_timestamp    order_approved_at  \
count         99441                    99441                99281   
unique            8                    98875                90733   
top       delivered      2018-04-11 10:48:14  2018-02-27 04:31:10   
freq          96478                        3                    9   

       order_delivered_carrier_date order_delivered_customer_date  \
count                         97658                         96476   
unique                        81018                         95664   
top             2018-05-09 15:48:00           2018-05-08

In [146]:
product_df = pd.read_csv("/content/products_dataset.csv",engine="python")
#Cetak info singkat dataset product_df
print(product_df.info())
print(f'Jumlah NaN values pada dataframe product_ yaitu {product_df.isna().sum()}')
print(f'Jumlah duplikasi pada dataframe product_df yaitu {product_df.duplicated().sum()}')


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB
None
Jumlah NaN values pada dataframe product_ yaitu product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty      

In [132]:
#Lakukan cleaning pada Data
product_df.dropna(how="any",inplace=True)
print(f"Jumlah NaN value sekarang adalah {product_df.isna().sum()}")


Jumlah NaN value sekarang adalah product_id                    0
product_category_name         0
product_name_lenght           0
product_description_lenght    0
product_photos_qty            0
product_weight_g              0
product_length_cm             0
product_height_cm             0
product_width_cm              0
dtype: int64


In [158]:
#Langsung ke seller dataset
seller_df = pd.read_csv("/content/sellers_dataset.csv",engine="python")
seller_df

seller_id  seller_zip_code_prefix  \
0     3442f8959a84dea7ee197c632cb2df15                   13023   
1     d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2     ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3     c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4     51a04a8a6bdcb23deccc82b0b80742cf                   12914   
...                                ...                     ...   
3090  98dddbc4601dd4443ca174359b237166                   87111   
3091  f8201cab383e484733266d1906e2fdfa                   88137   
3092  74871d19219c7d518d0090283e03c137                    4650   
3093  e603cf3fec55f8697c9059638d6c8eb5                   96080   
3094  9e25199f6ef7e7c347120ff175652c3b                   12051   

            seller_city seller_state  
0              campinas           SP  
1            mogi guacu           SP  
2        rio de janeiro           RJ  
3             sao paulo           SP  
4     braganca paulista           SP  
...                 ...          ...  
3090            sarandi           PR  
3091            palhoca           SC  
3092          sao paulo           SP  
3093            pelotas           RS  
3094            taubate           SP  

[3095 rows x 4 columns]

In [161]:
seller_df.describe()
print(f"Jumlah nilai terduplikasi : {seller_df.duplicated().sum()}")
print(f"Jumlah nilai NaN : {seller_df.isna().sum()}")

Jumlah nilai terduplikasi : 0
Jumlah nilai NaN : seller_id                 0
seller_zip_code_prefix    0
seller_city               0
seller_state              0
dtype: int64


In [162]:
seller_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   seller_id               3095 non-null   object
 1   seller_zip_code_prefix  3095 non-null   int64 
 2   seller_city             3095 non-null   object
 3   seller_state            3095 non-null   object
dtypes: int64(1), object(3)
memory usage: 96.8+ KB


In [170]:
#Geolocation_dataset
geolocation_df =pd.read_csv("/content/geolocation_dataset.csv")
print(geolocation_df.describe())
print(geolocation_df.info())

       geolocation_zip_code_prefix  geolocation_lat  geolocation_lng
count                 1.000163e+06     1.000163e+06     1.000163e+06
mean                  3.657417e+04    -2.117615e+01    -4.639054e+01
std                   3.054934e+04     5.715866e+00     4.269748e+00
min                   1.001000e+03    -3.660537e+01    -1.014668e+02
25%                   1.107500e+04    -2.360355e+01    -4.857317e+01
50%                   2.653000e+04    -2.291938e+01    -4.663788e+01
75%                   6.350400e+04    -1.997962e+01    -4.376771e+01
max                   9.999000e+04     4.506593e+01     1.211054e+02
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng       

In [171]:
#cek duplikasi sm NaN value
print(f"Jumlah nilai terduplikasi : {geolocation_df.duplicated().sum()}")
print(f"Jumlah nilai NaN : {geolocation_df.isna().sum()}")

#Data yang terduplikasi ada 261831, diwajarkan karena bisa jadi lokasi pengiriman barang berbeda tapi letak koordinat dan ketinggiannya sama.

Jumlah nilai terduplikasi : 261831
Jumlah nilai NaN : geolocation_zip_code_prefix    0
geolocation_lat                0
geolocation_lng                0
geolocation_city               0
geolocation_state              0
dtype: int64


In [206]:
order_reviews_df=pd.read_csv("/content/order_reviews_dataset.csv")
order_reviews_df

review_id                          order_id  \
0      7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1      80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2      228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3      e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4      f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   
...                                 ...                               ...   
99219  574ed12dd733e5fa530cfd4bbf39d7c9  2a8c23fee101d4d5662fa670396eb8da   
99220  f3897127253a9592a73be9bdfdf4ed7a  22ec9f0669f784db00fa86d035cf8602   
99221  b3de70c89b1510c4cd3d0649fd302472  55d4004744368f5571d1f590031933e4   
99222  1adeb9d84d72fe4e337617733eb85149  7725825d039fc1f0ceb7635e3f7d9206   
99223  efe49f1d6f951dd88b51e6ccd4cc548f  90531360ecb1eec2a1fbb265a0db0508   

       review_score review_comment_title  \
0                 4                  NaN   
1                 5                  NaN   
2                 5                  NaN   
3                 5                  NaN   
4                 5                  NaN   
...             ...                  ...   
99219             5                  NaN   
99220             5                  NaN   
99221             5                  NaN   
99222             4                  NaN   
99223             1                  NaN   

                                  review_comment_message review_creation_date  \
0                                                    NaN  2018-01-18 00:00:00   
1                                                    NaN  2018-03-10 00:00:00   
2                                                    NaN  2018-02-17 00:00:00   
3                  Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4      Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   
...                                                  ...                  ...   
99219                                                NaN  2018-07-07 00:00:00   
99220                                                NaN  2017-12-09 00:00:00   
99221  Excelente mochila, entrega super rápida. Super...  2018-03-22 00:00:00   
99222                                                NaN  2018-07-01 00:00:00   
99223  meu produto chegou e ja tenho que devolver, po...  2017-07-03 00:00:00   

      review_answer_timestamp  
0         2018-01-18 21:46:59  
1         2018-03-11 03:05:13  
2         2018-02-18 14:36:24  
3         2017-04-21 22:02:06  
4         2018-03-02 10:26:53  
...                       ...  
99219     2018-07-14 17:18:30  
99220     2017-12-11 20:06:42  
99221     2018-03-23 09:10:43  
99222     2018-07-02 12:59:13  
99223     2017-07-03 21:01:49  

[99224 rows x 7 columns]

In [211]:
print(f"Jumlah data terduplikasi : {order_reviews_df.duplicated().sum()}")
print(f"Jumlah NaN value:{order_reviews_df.isna().sum()}")
#Lakukan cleaning data pada review comment message dan comment_title
order_reviews_df.dropna(how="any",inplace=True)
#Ayo cek ulang
order_reviews_df.isna().sum()

Jumlah data terduplikasi : 0
Jumlah NaN value:review_id                  0
order_id                   0
review_score               0
review_comment_title       0
review_comment_message     0
review_creation_date       0
review_answer_timestamp    0
dtype: int64


review_id                  0
order_id                   0
review_score               0
review_comment_title       0
review_comment_message     0
review_creation_date       0
review_answer_timestamp    0
dtype: int64

# Exploratory Data Analysis

In [181]:
#kita akan mengecek data order yg sudah clean terlebih dahulu
customer_df.head(10)

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
5  879864dab9bc3047522c92c82e1212b8  4c93744516667ad3b8f1fb645a3116a4   
6  fd826e7cf63160e536e0908c76c3f441  addec96d2e059c80c30fe6871d30d177   
7  5e274e7a0c3809e14aba7ad5aae0d407  57b2a98a409812fe9618067b6b8ebe4f   
8  5adf08e34b2e993982a47070956c5c65  1175e95fb47ddff9de6b2b06188f7e0d   
9  4b7139f34592b3a31687243a302fa75b  9afe194fb833f79e300e37e580171f22   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP  
5                     89254         jaragua do sul             SC  
6                      4534              sao paulo             SP  
7                     35182                timoteo             MG  
8                     81560               curitiba             PR  
9                     30575         belo horizonte             MG

In [184]:
#golongkan berdasarkan customer_city dan customer_state
#Tiap user/pelanggan dari E-commerce ini pasti memiliki unique ID untuk membedakan. Maka kita gunakan nunique utk menghitung jumlah kode unique yg ada
customer_df.groupby(by=["customer_state","customer_city"]).agg({
    "customer_unique_id" : "nunique",
})


customer_unique_id
customer_state customer_city                              
AC             brasileia                                 1
               cruzeiro do sul                           3
               epitaciolandia                            1
               manoel urbano                             1
               porto acre                                1
...                                                    ...
TO             silvanopolis                              1
               sitio novo do tocantins                   1
               taguatinga                                3
               tocantinopolis                            7
               xambioa                                   3

[4310 rows x 1 columns]

In [193]:
customer_df.groupby(by="customer_city").customer_unique_id.nunique().sort_values(ascending=False)

customer_city
sao paulo         14984
rio de janeiro     6620
belo horizonte     2672
brasilia           2069
curitiba           1465
                  ...  
ibertioga             1
ibiajara              1
ibiara                1
ibicuitinga           1
lagoao                1
Name: customer_unique_id, Length: 4119, dtype: int64

Berdasarkan data ini diketahui bahwa pelanggan terbanyak berasal dari Sao Paulo, disusul dengan pelanggan dari Rio de Janeiro dan urutan ketiga Belo Horizonte. Sehingga **Pertanyaan 1** telah **terjawab**

In [187]:
customer_df.groupby(by='customer_state').customer_unique_id.nunique().sort_values(ascending = False)

customer_state
SP    40302
RJ    12384
MG    11259
RS     5277
PR     4882
SC     3534
BA     3277
DF     2075
ES     1964
GO     1952
PE     1609
CE     1313
PA      949
MT      876
MA      726
MS      694
PB      519
PI      482
RN      474
AL      401
SE      342
TO      273
RO      240
AM      143
AC       77
AP       67
RR       45
Name: customer_unique_id, dtype: int64

Diketahui bahwa pelanggan terbanyak berasal dari kode state SP dengan jumlah pelanggan 40302. Maka **Pertanyaan ke -2 (Negara bagian dengan Jumlah Pelanggan terbanyak)** telah **terjawab**

In [197]:
#Akan kita coba kaitkan dengan data order
orders_df


order_id                       customer_id  \
0      e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1      53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2      47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3      949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4      ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                 ...                               ...   
99436  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
99437  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
99438  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
99439  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
99440  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

      order_status order_purchase_timestamp    order_approved_at  \
0        delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1        delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2        delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3        delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4        delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   
...            ...                      ...                  ...   
99436    delivered      2017-03-09 09:54:05  2017-03-09 09:54:05   
99437    delivered      2018-02-06 12:58:58  2018-02-06 13:10:37   
99438    delivered      2017-08-27 14:46:43  2017-08-27 15:04:16   
99439    delivered      2018-01-08 21:28:27  2018-01-08 21:36:21   
99440    delivered      2018-03-08 20:57:30  2018-03-09 11:20:28   

      order_delivered_carrier_date order_delivered_customer_date  \
0              2017-10-04 19:55:00           2017-10-10 21:25:13   
1              2018-07-26 14:31:00           2018-08-07 15:27:45   
2              2018-08-08 13:50:00           2018-08-17 18:06:29   
3              2017-11-22 13:39:59           2017-12-02 00:28:42   
4              2018-02-14 19:46:34           2018-02-16 18:17:02   
...                            ...                           ...   
99436          2017-03-10 11:18:03           2017-03-17 15:08:01   
99437          2018-02-07 23:22:42           2018-02-28 17:37:56   
99438          2017-08-28 20:52:26           2017-09-21 11:24:17   
99439          2018-01-12 15:35:03           2018-01-25 23:32:54   
99440          2018-03-09 22:11:59           2018-03-16 13:08:30   

      order_estimated_delivery_date  
0               2017-10-18 00:00:00  
1               2018-08-13 00:00:00  
2               2018-09-04 00:00:00  
3               2017-12-15 00:00:00  
4               2018-02-26 00:00:00  
...                             ...  
99436           2017-03-28 00:00:00  
99437           2018-03-02 00:00:00  
99438           2017-09-27 00:00:00  
99439           2018-02-15 00:00:00  
99440           2018-04-03 00:00:00  

[96461 rows x 8 columns]

In [201]:
#Kita akan coba gabungkan antara data order dengan data customer. untuk melihat apakah barang yg dipesan pelanggan telah terkirim semua atau tidak
#They have something in common : customer_id
order_customer_df = pd.merge(
    left = orders_df,
    right = customer_df,
    how = "left",
    left_on = "customer_id",
    right_on = "customer_id"
)

order_customer_df

order_id                       customer_id  \
0      e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1      53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2      47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3      949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4      ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                 ...                               ...   
96456  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
96457  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
96458  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
96459  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
96460  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

      order_status order_purchase_timestamp    order_approved_at  \
0        delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1        delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2        delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3        delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4        delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   
...            ...                      ...                  ...   
96456    delivered      2017-03-09 09:54:05  2017-03-09 09:54:05   
96457    delivered      2018-02-06 12:58:58  2018-02-06 13:10:37   
96458    delivered      2017-08-27 14:46:43  2017-08-27 15:04:16   
96459    delivered      2018-01-08 21:28:27  2018-01-08 21:36:21   
96460    delivered      2018-03-08 20:57:30  2018-03-09 11:20:28   

      order_delivered_carrier_date order_delivered_customer_date  \
0              2017-10-04 19:55:00           2017-10-10 21:25:13   
1              2018-07-26 14:31:00           2018-08-07 15:27:45   
2              2018-08-08 13:50:00           2018-08-17 18:06:29   
3              2017-11-22 13:39:59           2017-12-02 00:28:42   
4              2018-02-14 19:46:34           2018-02-16 18:17:02   
...                            ...                           ...   
96456          2017-03-10 11:18:03           2017-03-17 15:08:01   
96457          2018-02-07 23:22:42           2018-02-28 17:37:56   
96458          2017-08-28 20:52:26           2017-09-21 11:24:17   
96459          2018-01-12 15:35:03           2018-01-25 23:32:54   
96460          2018-03-09 22:11:59           2018-03-16 13:08:30   

      order_estimated_delivery_date                customer_unique_id  \
0               2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
1               2018-08-13 00:00:00  af07308b275d755c9edb36a90c618231   
2               2018-09-04 00:00:00  3a653a41f6f9fc3d2a113cf8398680e8   
3               2017-12-15 00:00:00  7c142cf63193a1473d2e66489a9ae977   
4               2018-02-26 00:00:00  72632f0f9dd73dfee390c9b22eb56dd6   
...                             ...                               ...   
96456           2017-03-28 00:00:00  6359f309b166b0196dbf7ad2ac62bb5a   
96457           2018-03-02 00:00:00  da62f9e57a76d978d02ab5362c509660   
96458           2017-09-27 00:00:00  737520a9aad80b3fbbdad19b66b37b30   
96459           2018-02-15 00:00:00  5097a5312c8b157bb7be58ae360ef43c   
96460           2018-04-03 00:00:00  60350aa974b26ff12caad89e55993bd6   

       customer_zip_code_prefix            customer_city customer_state  
0                          3149                sao paulo             SP  
1                         47813                barreiras             BA  
2                         75265               vianopolis             GO  
3                         59296  sao goncalo do amarante             RN  
4                          9195              santo andre             SP  
...                         ...                      ...            ...  
96456                     12209      sao jose dos campos             SP  
964

In [204]:
#mari kita lihat, adakah barang yg tidak terkirim/tersasar?
order_customer_df.groupby(by = "order_status").customer_id.count()

order_status
canceled         6
delivered    96455
Name: customer_id, dtype: int64

In [205]:
#Ternyata ada yg dicancel, daerah mana itu?
order_customer_df.groupby(by=["order_status","customer_state"]).agg({
    "customer_id":"nunique"
})

customer_id
order_status customer_state             
canceled     MG                        1
             RJ                        3
             SC                        1
             SP                        1
delivered    AC                       80
             AL                      397
             AM                      145
             AP                       67
             BA                     3256
             CE                     1278
             DF                     2080
             ES                     1995
             GO                     1957
             MA                      716
             MG                    11351
             MS                      701
             MT                      886
             PA                      946
             PB                      517
             PE                     1593
             PI                      476
             PR                     4923
             RJ                    12348
             RN                      474
             RO                      243
             RR                       41
             RS                     5342
             SC                     3546
             SE                      335
             SP                    40488
             TO                      274

Berdasarkan data, ada beberapa barang yang dicancel dan user tersebut